File: OpenHarmony/developtools/hapsigner/hapsigntool_cpp/codesigning/sign/src/verify_code_signature.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "verify_code_signature.h"
#include "elf_sign_block.h"
#include "constant.h"

namespace OHOS {
namespace SignatureTools {
bool VerifyCodeSignature::VerifyHap(std::string file, int64_t offset, int64_t length,
                                    std::string fileFormat, std::string profileContent)
{
    std::string fileFromatTmp = fileFormat;
    std::transform(fileFromatTmp.begin(), fileFromatTmp.end(), fileFromatTmp.begin(), ::tolower);
    if (std::find(CodeSigning::SUPPORT_FILE_FORM.begin(), CodeSigning::SUPPORT_FILE_FORM.end(),
        fileFromatTmp) == CodeSigning::SUPPORT_FILE_FORM.end()) {
        SIGNATURE_TOOLS_LOGI("Not hap, hsp or hqf file, skip code signing verify, file type: %s",
            fileFromatTmp.c_str());
        return true;
    }
    // 1) generate CodeSignBlock
    std::pair<std::string, std::string> pairResult = HapUtils::ParseAppIdentifier(profileContent);
    CodeSignBlock csb;
    bool generateFlag = GenerateCodeSignBlock(file, offset, length, csb);
    if (!generateFlag) {
        SIGNATURE_TOOLS_LOGE("VerifyHap GenerateCodeSignBlock failed, file: %s", file.c_str());
        return false;
    }
    // 2) verify hap
    bool verifyFlag = VerifyCodeSign(file, pairResult, csb);
    if (!verifyFlag) {
        SIGNATURE_TOOLS_LOGE("VerifyHap VerifyCodeSign failed");
        return false;
    }
    return true;
}

bool VerifyCodeSignature::VerifyElf(std::string file, int64_t offset, int64_t length,
                                    std::string fileFormat, std::string profileContent)
{
    std::transform(fileFormat.begin(), fileFormat.end(), fileFormat.begin(), ::tolower);
    if (ELF != fileFormat) {
        SIGNATURE_TOOLS_LOGI("Not elf file, skip code signing verify, file type: %s", fileFormat.c_str());
        return true;
    }
    // 1) parse sign block to ElfSignBlock object
    std::ifstream signedElf(file, std::ios::binary);
    if (!signedElf.is_open()) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "open file: " + file + "failed");
        return false;
    }
    signedElf.seekg(offset, std::ios::beg);
    std::vector<char> codeSignBlockBytes(length);
    signedElf.read(codeSignBlockBytes.data(), codeSignBlockBytes.size());
    signedElf.close();
    ElfSignBlock elfSignBlock;
    bool arrFlag = ElfSignBlock::FromByteArray(*(std::vector<int8_t> *) & codeSignBlockBytes, elfSignBlock);
    if (!arrFlag) {
        SIGNATURE_TOOLS_LOGE("parse sign block to ElfCodeSignBlock object failed");
        return false;
    }
    // 2) verify file data
    int32_t paddingSize = ElfSignBlock::ComputeMerkleTreePaddingLength(offset);
    std::vector<int8_t> &merkleTreeWithPadding = elfSignBlock.GetMerkleTreeWithPadding();
    std::vector<int8_t> merkleTree;
    merkleTree.insert(merkleTree.end(), merkleTreeWithPadding.begin() + paddingSize, merkleTreeWithPadding.end());
    std::ifstream elf(file, std::ios::binary);
    bool verifySingleFileFlag = VerifySingleFile(elf, elfSignBlock.GetDataSize(), elfSignBlock.GetSignature(),
        elfSignBlock.GetTreeOffset(), merkleTree);
    if (!verifySingleFileFlag) {
        SIGNATURE_TOOLS_LOGE("verify elf file data failed");
        elf.close();
        return false;
    }
    elf.close();
    // 3) check ownerID
    if (!profileContent.empty()) {
        std::pair<std::string, std::string> pairResult = HapUtils::ParseAppIdentifier(profileContent);
        std::vector<int8_t>& signature = elfSignBlock.GetSignature();
        std::string signatureStr(signature.begin(), signature.end());
        bool checkOwnerIDFlag = CmsUtils::CheckOwnerID(signatureStr, pairResult.first, pairResult.second);
        if (!checkOwnerIDFlag) {
            SIGNATURE_TOOLS_LOGE("elf check owner id failed");
            return false;
        }
    }
    return true;
}

bool VerifyCodeSignature::VerifyNativeLib(CodeSignBlock& csb, std::string& file, unzFile& zFile,
                                          std::pair<std::string, std::string>& pairResult)
{
    UnzipHandleParam param(csb, pairResult, false);
    CodeSigning codeSigning;
    bool nativeFlag = codeSigning.GetNativeEntriesFromHap(file, param);
    if (!nativeFlag) {
        SIGNATURE_TOOLS_LOGE("verify native libs failed, file: %s", file.c_str());
        return false;
    }
    return true;
}

bool VerifyCodeSignature::VerifyCodeSign(std::string file, std::pair<std::string, std::string>& pairResult,
                                         CodeSignBlock& csb)
{
    // 1) verify hap
    std::ifstream hap(file, std::ios::binary);
    if (!hap.is_open()) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "open file: " + file + "failed");
        return false;
    }
    int64_t dataSize = csb.GetHapInfoSegment().GetSignInfo().GetDataSize();
    std::vector<int8_t> signature = csb.GetHapInfoSegment().GetSignInfo().GetSignature();
    Extension* ext = csb.GetHapInfoSegment().GetSignInfo()
        .GetExtensionByType(MerkleTreeExtension::MERKLE_TREE_INLINED);
    MerkleTreeExtension* mte = new MerkleTreeExtension(0, 0, std::vector<int8_t>());
    if (mte == nullptr) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "system failed to allocate memory for MerkleTreeExtension");
        return false;
    }
    if (ext != nullptr) {
        delete mte;
        mte = static_cast<MerkleTreeExtension*>(ext);
    } else {
        std::shared_ptr<MerkleTreeExtension> merkleTreeExt(mte);
    }
    // temporary: merkle tree offset set to zero, change to merkleTreeOffset
    bool verifyFlag = VerifySingleFile(hap, dataSize, signature, mte->GetMerkleTreeOffset(),
        csb.GetOneMerkleTreeByFileName(CodeSigning::HAP_SIGNATURE_ENTRY_NAME));
    if (!verifyFlag) {
        return false;
    }
    hap.close();
    std::string signature_(signature.begin(), signature.end());
    bool checkOwnerIDFlag = CmsUtils::CheckOwnerID(signature_, pairResult.first, pairResult.second);
    if (!checkOwnerIDFlag) {
        return false;
    }
    // 2) verify native libs
    unzFile zFile = unzOpen(file.c_str());
    if (zFile == NULL) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "unzOpen file: " + file + "failed");
        return false;
    }
    bool verifyNativeLibFlag = VerifyNativeLib(csb, file, zFile, pairResult);
    if (!verifyNativeLibFlag) {
        return false;
    }
    unzClose(zFile);
    return true;
}

bool VerifyCodeSignature::VerifySingleFile(std::istream& input, int64_t length,
                                           std::vector<int8_t> signature,
                                           int64_t merkleTreeOffset,
                                           std::vector<int8_t> inMerkleTreeBytes)
{
    std::pair<std::vector<int8_t>, std::vector<int8_t>> pairResult
        = GenerateFsVerityDigest(input, length, merkleTreeOffset);
    std::vector<int8_t> generatedMerkleTreeBytes = pairResult.second;
    // For native libs, inMerkleTreeBytes is null, skip check here
    bool verifyFlag = (!inMerkleTreeBytes.empty()) && !AreVectorsEqual(inMerkleTreeBytes, generatedMerkleTreeBytes);
    if (verifyFlag) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "verify merkle tree bytes failed");
        return false;
    }
    return CmsUtils::VerifySignDataWithUnsignedDataDigest(pairResult.first, signature);
}

bool VerifyCodeSignature::GenerateCodeSignBlock(const std::string& file, int64_t offset,
                                                int64_t length, CodeSignBlock& csb)
{
    // 1) parse sign block to CodeSignBlock object
    std::ifstream signedHap(file, std::ios::binary);
    if (!signedHap.is_open()) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "open file: " + file + "failed");
        return false;
    }
    int32_t fileReadOffset = 0;
    // 1.1) skip data part, but fileReadOffset remains at start(0)
    signedHap.seekg(offset, std::ios::beg);
    // 1.2) parse codeSignBlockHeader
    std::vector<char> codeSignBlockHeaderByteArray(CodeSignBlockHeader::Size());
    signedHap.read(codeSignBlockHeaderByteArray.data(), codeSignBlockHeaderByteArray.size());
    fileReadOffset += signedHap.gcount();
    CodeSignBlockHeader* pCodeSignBlockHeader =
        CodeSignBlockHeader::FromByteArray(*(std::vector<int8_t>*) &codeSignBlockHeaderByteArray);
    if (pCodeSignBlockHeader == nullptr) {
        SIGNATURE_TOOLS_LOGE("Invalid code Sign block header");
        signedHap.close();
        return false;
    }
    std::shared_ptr<CodeSignBlockHeader> codeSignBlockHeader(pCodeSignBlockHeader);
    csb.SetCodeSignBlockHeader(*codeSignBlockHeader);
    int64_t headerBlockSize = csb.GetCodeSignBlockHeader().GetBlockSize();
    if (headerBlockSize != length) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR,
            "Invalid code Sign block size of setCodeSignBlockHeader");
        signedHap.close();
        return false;
    }
    // 2) parse segment headers
    for (int i = 0; i < csb.GetCodeSignBlockHeader().GetSegmentNum(); i++) {
        std::vector<char> segmentHeaderByteArray(SegmentHeader::SEGMENT_HEADER_LENGTH);
        signedHap.read(segmentHeaderByteArray.data(), segmentHeaderByteArray.size());
        fileReadOffset += signedHap.gcount();
        csb.AddToSegmentList(*SegmentHeader::FromByteArray(*(std::vector<int8_t> *) &segmentHeaderByteArray));
    }
    // compute merkle tree offset by alignment, based on file start
    int64_t computedTreeOffset = GetAlignmentAddr(CodeSignBlock::PAGE_SIZE_4K, fileReadOffset + offset);
    // skip zero padding before merkle tree, adds zero padding length to fileReadOffset
    signedHap.seekg(computedTreeOffset - offset - fileReadOffset, std::ios::cur);
    fileReadOffset += computedTreeOffset - offset - fileReadOffset;
    bool parseMerkleTreeFlag = ParseMerkleTree(csb, fileReadOffset, signedHap, computedTreeOffset);
    if (!parseMerkleTreeFlag) {
        SIGNATURE_TOOLS_LOGE("GenerateCodeSignBlock parse merkle tree failed.");
        return false;
    }
    signedHap.close();
    return true;
}

bool VerifyCodeSignature::ParseSegmentHead(CodeSignBlock& csb, std::ifstream& signedHap,
    std::vector<char> &merkleTreeBytes, int32_t &fileReadOffset)
{
    for (auto& segmentHeader : csb.GetSegmentHeaderList()) {
        if (fileReadOffset > segmentHeader.GetSegmentOffset()) {
            PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "Invaild offset of merkle tree and segment header");
            return false;
        }
        // get merkle tree bytes
        if (fileReadOffset < segmentHeader.GetSegmentOffset()) {
            merkleTreeBytes.resize(segmentHeader.GetSegmentOffset() - fileReadOffset);
            signedHap.read(merkleTreeBytes.data(), merkleTreeBytes.size());
            fileReadOffset += signedHap.gcount();
        }
        std::vector<char> sh(segmentHeader.GetSegmentSize(), 0);
        signedHap.read(sh.data(), sh.size());
        fileReadOffset += signedHap.gcount();
        int32_t type = segmentHeader.GetType();
        if (type == SegmentHeader::CSB_FSVERITY_INFO_SEG) {
            // 3) parse fs-verity info segment
            FsVerityInfoSegment fsVerityInfoSegment =
                FsVerityInfoSegment::FromByteArray(*(std::vector<int8_t> *) & sh);
            csb.SetFsVerityInfoSegment(fsVerityInfoSegment);
        } else if (type == SegmentHeader::CSB_HAP_META_SEG) {
            // 4) parse hap info segment
            HapInfoSegment hapInfoSegment = HapInfoSegment::FromByteArray(*(std::vector<int8_t> *) & sh);
            csb.SetHapInfoSegment(hapInfoSegment);
        } else if (type == SegmentHeader::CSB_NATIVE_LIB_INFO_SEG) {
            // 5) parse so info segment
            NativeLibInfoSegment nativeLibInfoSegment =
                NativeLibInfoSegment::FromByteArray(*(std::vector<int8_t> *) & sh);
            csb.SetSoInfoSegment(nativeLibInfoSegment);
        }
    }

    return true;
}

bool VerifyCodeSignature::ParseMerkleTree(CodeSignBlock& csb, int32_t readOffset, std::ifstream& signedHap,
                                          int64_t computedTreeOffset)
{
    std::vector<char> merkleTreeBytes;
    int32_t fileReadOffset = readOffset;
    bool parseFlag = ParseSegmentHead(csb, signedHap, merkleTreeBytes, fileReadOffset);
    if (!parseFlag) {
        SIGNATURE_TOOLS_LOGE("ParsesegmentHead failed");
        return false;
    }
    int32_t blockSize = csb.GetCodeSignBlockHeader().GetBlockSize();
    if (fileReadOffset != blockSize) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR,
            "Invalid blockSize of getCodeSignBlockHeader");
        return false;
    }
    // parse merkle tree
    Extension* extension = csb.GetHapInfoSegment().GetSignInfo().
        GetExtensionByType(MerkleTreeExtension::MERKLE_TREE_INLINED);
    if (!extension) {
        SIGNATURE_TOOLS_LOGE("Missing merkleTreeExtension in verifycation");
        return false;
    }
    MerkleTreeExtension* mte = static_cast<MerkleTreeExtension*>(extension);
    if (mte != nullptr) {
        bool merkleTreeFlag = computedTreeOffset != mte->GetMerkleTreeOffset() ||
            merkleTreeBytes.size() != mte->GetMerkleTreeSize();
        if (merkleTreeFlag) {
            PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "Invalid merkle tree offset or tree size");
            return false;
        }
        csb.AddOneMerkleTree(CodeSigning::HAP_SIGNATURE_ENTRY_NAME, *(std::vector<int8_t> *) &merkleTreeBytes);
    }
    return true;
}

int64_t VerifyCodeSignature::GetAlignmentAddr(int64_t alignment, int64_t input)
{
    if (alignment == 0) {
        return input;
    }
    int64_t residual = input % alignment;
    if (residual == 0) {
        return input;
    } else {
        return input + (alignment - residual);
    }
}

std::pair<std::vector<int8_t>, std::vector<int8_t>> VerifyCodeSignature::GenerateFsVerityDigest(
    std::istream& inputStream, int64_t size, int64_t merkleTreeOffset)
{
    std::unique_ptr<FsVerityGenerator> fsVerityGenerator = std::make_unique<FsVerityGenerator>();
    fsVerityGenerator->GenerateFsVerityDigest(inputStream, size, merkleTreeOffset);
    return std::make_pair(fsVerityGenerator->GetFsVerityDigest(), fsVerityGenerator->GetTreeBytes());
}

bool VerifyCodeSignature::AreVectorsEqual(const std::vector<int8_t>& vec1, const std::vector<int8_t>& vec2)
{
    if (vec1.size() != vec2.size()) {
        return false;
    }
    for (int64_t i = 0; i < vec1.size(); i++) {
        if (vec1[i] != vec2[i]) {
            return false;
        }
    }
    return true;
}
} // namespace SignatureTools
} // namespace OHOS

File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/codesigning/sign/VerifyCodeSignature.java
/*
 * Copyright (c) 2023-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.codesigning.sign;

import com.ohos.hapsigntool.codesigning.datastructure.CodeSignBlock;
import com.ohos.hapsigntool.codesigning.datastructure.CodeSignBlockHeader;
import com.ohos.hapsigntool.codesigning.datastructure.ElfSignBlock;
import com.ohos.hapsigntool.codesigning.datastructure.Extension;
import com.ohos.hapsigntool.codesigning.datastructure.FsVerityInfoSegment;
import com.ohos.hapsigntool.codesigning.datastructure.HapInfoSegment;
import com.ohos.hapsigntool.codesigning.datastructure.MerkleTreeExtension;
import com.ohos.hapsigntool.codesigning.datastructure.NativeLibInfoSegment;
import com.ohos.hapsigntool.codesigning.datastructure.SegmentHeader;
import com.ohos.hapsigntool.codesigning.datastructure.SignInfo;
import com.ohos.hapsigntool.codesigning.exception.FsVerityDigestException;
import com.ohos.hapsigntool.codesigning.exception.PageInfoException;
import com.ohos.hapsigntool.codesigning.exception.VerifyCodeSignException;
import com.ohos.hapsigntool.codesigning.fsverity.FsVerityGenerator;
import com.ohos.hapsigntool.codesigning.utils.CmsUtils;
import com.ohos.hapsigntool.codesigning.utils.HapUtils;
import com.ohos.hapsigntool.entity.Pair;
import com.ohos.hapsigntool.error.ProfileException;
import com.ohos.hapsigntool.utils.LogUtils;
import com.ohos.hapsigntool.utils.StringUtils;

import org.bouncycastle.asn1.ASN1ObjectIdentifier;
import org.bouncycastle.asn1.cms.Attribute;
import org.bouncycastle.asn1.cms.AttributeTable;
import org.bouncycastle.cms.CMSException;
import org.bouncycastle.cms.CMSSignedData;
import org.bouncycastle.cms.SignerInformation;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import java.util.zip.ZipInputStream;

/**
 * Verify code signature given a file with code sign block
 *
 * @since 2023/09/08
 */
public class VerifyCodeSignature {
    private static final LogUtils LOGGER = new LogUtils(VerifyCodeSignature.class);

    private static void checkOwnerID(byte[] signature, String profileOwnerID, String profileType)
        throws CMSException, VerifyCodeSignException {
        String ownerID = profileOwnerID;
        // if profileType is debug, check the app-id in signature, should be null or DEBUG_LIB_ID
        if ("debug".equals(profileType)) {
            ownerID = HapUtils.HAP_DEBUG_OWNER_ID;
        }
        checkSignatureOwnerID(ownerID, signature, profileType);
    }

    private static void checkHnpOwnerID(byte[] signature, String profileOwnerID, String profileType, String hnpType)
        throws CMSException, VerifyCodeSignException {
        String ownerID = profileOwnerID;
        // if profileType is debug, check the app-id in signature, should be null or DEBUG_LIB_ID
        if ("debug".equals(profileType)) {
            ownerID = HapUtils.HAP_DEBUG_OWNER_ID;
        } else if ("release".equals(profileType)) {
            if ("public".equals(hnpType)) {
                ownerID = HapUtils.HAP_SHARED_OWNER_ID;
            }
        }
        checkSignatureOwnerID(ownerID, signature, profileType);
    }

    private static void checkSignatureOwnerID(String ownerID, byte[] signature, String profileType)
        throws CMSException, VerifyCodeSignException {
        CMSSignedData cmsSignedData = new CMSSignedData(signature);
        Collection<SignerInformation> signers = cmsSignedData.getSignerInfos().getSigners();
        for (SignerInformation signer : signers) {
            AttributeTable attrTable = signer.getSignedAttributes();
            Attribute attr = attrTable.get(new ASN1ObjectIdentifier(BcSignedDataGenerator.SIGNER_OID));
            // if app-id is null, if profileType is debug, it's ok.
            if (attr == null) {
                if ("debug".equals(profileType)) {
                    continue;
                }
                if (ownerID == null) {
                    continue;
                } else {
                    throw new VerifyCodeSignException("app-identifier is not in the signature");
                }
            }
            if (ownerID == null) {
                throw new VerifyCodeSignException("app-identifier in profile is null, but is not null in signature");
            }
            // if app-id in signature exists, it should be equal to the app-id in profile.
            String resultOwnerID = attr.getAttrValues().getObjectAt(0).toString();
            if (!ownerID.equals(resultOwnerID)) {
                throw new VerifyCodeSignException("app-identifier in signature is invalid");
            }
        }
    }

    /**
     * Verify a signed elf's signature
     *
     * @param file       signed elf file
     * @param offset     start position of code sign block based on the start of the elf file
     * @param length     byte size of code sign block
     * @param fileFormat elf
     * @param profileContent profile json string
     * @return true if signature verify succeed and false otherwise
     * @throws IOException             If an input or output exception occurred
     * @throws VerifyCodeSignException parsing result invalid
     * @throws FsVerityDigestException if fs-verity digest generation failed
     * @throws CMSException            if signature verify failed
     * @throws ProfileException        if verify profile failed
     */
    public static boolean verifyElf(File file, long offset, long length, String fileFormat, String profileContent)
            throws IOException, VerifyCodeSignException, FsVerityDigestException, CMSException, ProfileException {
        if (!CodeSigning.SUPPORT_BIN_FILE_FORM.equalsIgnoreCase(fileFormat)) {
            LOGGER.info("Not elf file, skip code signing verify");
            return true;
        }
        // 1) parse sign block to ElfCodeSignBlock object
        ElfSignBlock elfSignBlock;
        try (FileInputStream signedElf = new FileInputStream(file)) {
            byte[] codeSignBlockBytes = new byte[(int) length];
            signedElf.skip(offset);
            signedElf.read(codeSignBlockBytes);
            elfSignBlock = ElfSignBlock.fromByteArray(codeSignBlockBytes);
        }
        // 2) verify file data
        try (FileInputStream signedElf = new FileInputStream(file)) {
            int paddingSize = ElfSignBlock.computeMerkleTreePaddingLength(offset);
            byte[] merkleTreeWithPadding = elfSignBlock.getMerkleTreeWithPadding();
            byte[] merkleTree = Arrays.copyOfRange(merkleTreeWithPadding, paddingSize, merkleTreeWithPadding.length);
            verifySingleFile(signedElf, elfSignBlock.getDataSize(), elfSignBlock.getSignature(),
                elfSignBlock.getTreeOffset(), merkleTree);
        }
        if (profileContent != null) {
            Pair<String, String> pairResult = HapUtils.parseAppIdentifier(profileContent);
            checkOwnerID(elfSignBlock.getSignature(), pairResult.getFirst(), pairResult.getSecond());
        }
        return true;
    }

    /**
     * Verify a signed hap's signature
     *
     * @param file       signed hap file
     * @param offset     start position of code sign block based on the start of the hap file
     * @param length     byte size of code sign block
     * @param fileFormat hap or hqf or hsp, etc.
     * @param profileContent profile of the hap
     * @return true if signature verify succeed and false otherwise
     * @throws IOException             If an input or output exception occurred
     * @throws VerifyCodeSignException parsing result invalid
     * @throws FsVerityDigestException if fs-verity digest generation failed
     * @throws CMSException            if signature verify failed
     * @throws ProfileException        profile of the hap failed
     */
    public static boolean verifyHap(File file, long offset, long length, String fileFormat, String profileContent)
            throws IOException, VerifyCodeSignException, FsVerityDigestException, CMSException, ProfileException {
        if (!StringUtils.containsIgnoreCase(CodeSigning.SUPPORT_FILE_FORM, fileFormat)) {
            LOGGER.info("Not hap, hsp or hqf file, skip code signing verify");
            return true;
        }
        Pair<String, String> pairResult = HapUtils.parseAppIdentifier(profileContent);

        CodeSignBlock csb = generateCodeSignBlock(file, offset, length);
        // 2) verify hap
        try (FileInputStream hap = new FileInputStream(file)) {
            long dataSize = csb.getHapInfoSegment().getSignInfo().getDataSize();
            byte[] signature = csb.getHapInfoSegment().getSignInfo().getSignature();
            Extension extension = csb.getHapInfoSegment()
                .getSignInfo()
                .getExtensionByType(MerkleTreeExtension.MERKLE_TREE_INLINED);
            MerkleTreeExtension mte = new MerkleTreeExtension(0, 0, null);
            if (extension instanceof MerkleTreeExtension) {
                mte = (MerkleTreeExtension) extension;
            }
            // temporary: merkle tree offset set to zero, change to merkleTreeOffset
            verifySingleFile(hap, dataSize, signature, mte.getMerkleTreeOffset(),
                csb.getOneMerkleTreeByFileName(CodeSigning.HAP_SIGNATURE_ENTRY_NAME));
            checkOwnerID(signature, pairResult.getFirst(), pairResult.getSecond());
        }
        // 3) verify native libs
        verifyLibs(file, csb, pairResult);
        return true;
    }

    private static void verifyLibs(File file, CodeSignBlock csb, Pair<String, String> pairResult)
        throws IOException, FsVerityDigestException, VerifyCodeSignException, CMSException, ProfileException {
        try (JarFile inputJar = new JarFile(file, false)) {
            Map<String, SignInfo> hnpLibSignInfoMap = new HashMap<>();
            Set<String> hnpEntryNames = new HashSet<>();
            for (int i = 0; i < csb.getSoInfoSegment().getSectionNum(); i++) {
                String entryName = csb.getSoInfoSegment().getFileNameList().get(i);
                SignInfo signInfo = csb.getSoInfoSegment().getSignInfoList().get(i);
                if (entryName.contains("!/")) {
                    String[] filePath = entryName.split("!/");
                    hnpEntryNames.add(filePath[0]);
                    hnpLibSignInfoMap.put(entryName, signInfo);
                } else {
                    LOGGER.info("verify lib: {}", entryName);
                    verifyHapLib(inputJar, entryName, signInfo, pairResult);
                }
            }
            if (hnpEntryNames.isEmpty()) {
                // not exists hnp lib
                return;
            }
            // get module.json
            Map<String, String> hnpTypeMap = HapUtils.getHnpsFromJson(inputJar);
            for (String hnpEntryName : hnpEntryNames) {
                verifyHnpLib(inputJar, hnpEntryName, hnpLibSignInfoMap, hnpTypeMap, pairResult);
            }
        }
    }

    private static void verifyHapLib(JarFile inputJar, String entryName, SignInfo signInfo,
        Pair<String, String> pairResult)
        throws IOException, FsVerityDigestException, VerifyCodeSignException, CMSException {

        JarEntry entry = inputJar.getJarEntry(entryName);
        if (entry.getSize() != signInfo.getDataSize()) {
            throw new VerifyCodeSignException(
                String.format(Locale.ROOT, "Invalid dataSize of native lib %s", entryName));
        }
        byte[] entrySig = signInfo.getSignature();
        try (InputStream entryInputStream = inputJar.getInputStream(entry)) {
            // temporary merkleTreeOffset 0
            verifySingleFile(entryInputStream, entry.getSize(), entrySig, 0, null);
            checkOwnerID(entrySig, pairResult.getFirst(), pairResult.getSecond());
        }
    }

    private static void verifyHnpLib(JarFile inputJar, String hnpEntryName, Map<String, SignInfo> hnpLibSignInfoMap,
        Map<String, String> hnpTypeMap, Pair<String, String> pairResult)
        throws IOException, FsVerityDigestException, VerifyCodeSignException, CMSException {
        JarEntry hnpEntry = inputJar.getJarEntry(hnpEntryName);
        try (InputStream inputStream = inputJar.getInputStream(hnpEntry);
            ZipInputStream hnpInputStream = new ZipInputStream(inputStream)) {
            String hnpFileName = HapUtils.parseHnpPath(hnpEntryName);
            if (!hnpTypeMap.containsKey(hnpFileName)) {
                throw new VerifyCodeSignException("hnp should be described in module.json");
            }
            String hnpType = hnpTypeMap.get(hnpFileName);
            java.util.zip.ZipEntry libEntry = null;
            while ((libEntry = hnpInputStream.getNextEntry()) != null) {
                String libPath = hnpEntry.getName() + "!/" + libEntry.getName();
                if (!hnpLibSignInfoMap.containsKey(libPath)) {
                    continue;
                }
                LOGGER.info("verify lib: {}", libPath);
                SignInfo signInfo = hnpLibSignInfoMap.get(libPath);
                byte[] entrySig = signInfo.getSignature();
                long dataSize = signInfo.getDataSize();
                verifySingleFile(hnpInputStream, dataSize, entrySig, 0, null);
                checkHnpOwnerID(entrySig, pairResult.getFirst(), pairResult.getSecond(), hnpType);
                hnpInputStream.closeEntry();
            }
        }
    }

    private static CodeSignBlock generateCodeSignBlock(File file, long offset, long length)
        throws IOException, VerifyCodeSignException {
        CodeSignBlock csb = new CodeSignBlock();
        // 1) parse sign block to CodeSignBlock object
        try (FileInputStream signedHap = new FileInputStream(file)) {
            int fileReadOffset = 0;
            // 0) skip data part, but fileReadOffset remains at start(0)
            signedHap.skip(offset);
            // 1) parse codeSignBlockHeader
            byte[] codeSignBlockHeaderByteArray = new byte[CodeSignBlockHeader.size()];
            fileReadOffset += signedHap.read(codeSignBlockHeaderByteArray);
            csb.setCodeSignBlockHeader(CodeSignBlockHeader.fromByteArray(codeSignBlockHeaderByteArray));
            if (csb.getCodeSignBlockHeader().getBlockSize() != length) {
                throw new VerifyCodeSignException("Invalid code Sign block size of setCodeSignBlockHeader");
            }
            // 2) parse segment headers
            for (int i = 0; i < csb.getCodeSignBlockHeader().getSegmentNum(); i++) {
                byte[] segmentHeaderByteArray = new byte[SegmentHeader.SEGMENT_HEADER_LENGTH];
                fileReadOffset += signedHap.read(segmentHeaderByteArray);
                csb.addToSegmentList(SegmentHeader.fromByteArray(segmentHeaderByteArray));
            }
            // compute merkle tree offset by alignment, based on file start
            long computedTreeOffset = getAlignmentAddr(CodeSignBlock.PAGE_SIZE_4K, fileReadOffset + offset);
            // skip zero padding before merkle tree, adds zero padding length to fileReadOffset
            fileReadOffset += signedHap.skip(computedTreeOffset - offset - fileReadOffset);
            parseMerkleTree(csb, fileReadOffset, signedHap, computedTreeOffset);
        }
        return csb;
    }

    private static void parseMerkleTree(CodeSignBlock csb, int readOffset, FileInputStream signedHap,
        long computedTreeOffset) throws VerifyCodeSignException, IOException {
        // check segment offset and segment size
        byte[] merkleTreeBytes = new byte[0];
        int fileReadOffset = readOffset;
        for (SegmentHeader segmentHeader : csb.getSegmentHeaderList()) {
            if (fileReadOffset > segmentHeader.getSegmentOffset()) {
                throw new VerifyCodeSignException("Invaild offset of merkle tree and segment header");
            }
            // get merkle tree bytes
            if (fileReadOffset < segmentHeader.getSegmentOffset()) {
                merkleTreeBytes = new byte[segmentHeader.getSegmentOffset() - fileReadOffset];
                fileReadOffset += signedHap.read(merkleTreeBytes);
            }
            byte[] sh = new byte[segmentHeader.getSegmentSize()];
            fileReadOffset += signedHap.read(sh);
            if (segmentHeader.getType() == SegmentHeader.CSB_FSVERITY_INFO_SEG) {
                // 3) parse fs-verity info segment
                csb.setFsVerityInfoSegment(FsVerityInfoSegment.fromByteArray(sh));
            } else if (segmentHeader.getType() == SegmentHeader.CSB_HAP_META_SEG) {
                // 4) parse hap info segment
                csb.setHapInfoSegment(HapInfoSegment.fromByteArray(sh));
            } else if (segmentHeader.getType() == SegmentHeader.CSB_NATIVE_LIB_INFO_SEG) {
                // 5) parse so info segment
                csb.setSoInfoSegment(NativeLibInfoSegment.fromByteArray(sh));
            }
        }
        if (fileReadOffset != csb.getCodeSignBlockHeader().getBlockSize()) {
            throw new VerifyCodeSignException("Invalid blockSize of getCodeSignBlockHeader");
        }
        // parse merkle tree
        Extension extension = csb.getHapInfoSegment()
            .getSignInfo()
            .getExtensionByType(MerkleTreeExtension.MERKLE_TREE_INLINED);
        if (extension == null) {
            throw new VerifyCodeSignException("Missing merkleTreeExtension in verifycation");
        }
        if (extension instanceof MerkleTreeExtension) {
            MerkleTreeExtension mte = (MerkleTreeExtension) extension;
            if (computedTreeOffset != mte.getMerkleTreeOffset()) {
                throw new VerifyCodeSignException("Invalid merkle tree offset");
            }
            if (merkleTreeBytes.length != mte.getMerkleTreeSize()) {
                throw new VerifyCodeSignException("Invalid merkle tree size");
            }
            csb.addOneMerkleTree(CodeSigning.HAP_SIGNATURE_ENTRY_NAME, merkleTreeBytes);
        }
    }

    private static long getAlignmentAddr(long alignment, long input) {
        long residual = input % alignment;
        if (residual == 0) {
            return input;
        } else {
            return input + (alignment - residual);
        }
    }

    /**
     * Verifies the signature of a given file with its signature in pkcs#7 format
     *
     * @param input             file being verified in InputStream representation
     * @param length            size of signed data in the file
     * @param signature         byte array of signature in pkcs#7 format
     * @param merkleTreeOffset  merkle tree offset based on file start
     * @param inMerkleTreeBytes merkle tree raw bytes
     * @throws FsVerityDigestException if fs-verity digest generation failed
     * @throws CMSException            if signature verify failed
     * @throws VerifyCodeSignException parsing code sign block error
     */
    public static void verifySingleFile(InputStream input, long length, byte[] signature, long merkleTreeOffset,
        byte[] inMerkleTreeBytes) throws FsVerityDigestException, CMSException, VerifyCodeSignException {
        Pair<byte[], byte[]> pairResult = null;
        try {
            pairResult = generateFsVerityDigest(input, length, merkleTreeOffset);
        } catch (PageInfoException e) {
            throw new VerifyCodeSignException(e.getMessage());
        }
        byte[] generatedMerkleTreeBytes = pairResult.getSecond();
        if (generatedMerkleTreeBytes == null) {
            generatedMerkleTreeBytes = new byte[0];
        }
        // For native libs, inMerkleTreeBytes is null, skip check here
        if ((inMerkleTreeBytes != null) && !Arrays.equals(inMerkleTreeBytes, generatedMerkleTreeBytes)) {
            throw new VerifyCodeSignException("verify merkle tree bytes failed");
        }
        CmsUtils.verifySignDataWithUnsignedDataDigest(pairResult.getFirst(), signature);
    }

    private static Pair<byte[], byte[]> generateFsVerityDigest(InputStream inputStream, long size,
        long merkleTreeOffset) throws FsVerityDigestException, PageInfoException {
        FsVerityGenerator fsVerityGenerator = new FsVerityGenerator();
        fsVerityGenerator.generateFsVerityDigest(inputStream, size, merkleTreeOffset);
        return Pair.create(fsVerityGenerator.getFsVerityDigest(), fsVerityGenerator.getTreeBytes());
    }
}
